### ISSUE TITLE
Required `Param` requires `None` to be explicitly passed as default

### ISSUE THREAD

**Issue by aamster** [apache/airflow#28940]
### Apache Airflow version

2.5.0

### What happened

I have a dag like

```
import datetime

from airflow.models import Param
from airflow.models.dag import dag


@dag(
    dag_id="foo",
    schedule=None,
    params={'month': Param()},
    start_date=datetime.datetime.now()
)
def foo():
    pass


foo()

```

I am trying to pass this param via CLI. I tried passing it like 
```
airflow dags test foo --conf '{"month": "2022-01"}'
```

gives
```
Traceback (most recent call last):
  File "/Users/adam.amster/venvs/airflow/lib/python3.10/site-packages/airflow/models/dagbag.py", line 437, in _process_modules
    dag.validate()
  File "/Users/adam.amster/venvs/airflow/lib/python3.10/site-packages/airflow/models/dag.py", line 667, in validate
    self.params.validate()
  File "/Users/adam.amster/venvs/airflow/lib/python3.10/site-packages/airflow/models/param.py", line 230, in validate
    raise ParamValidationError(f"Invalid input for param {k}: {ve}") from None
airflow.exceptions.ParamValidationError: Invalid input for param month: No value passed and Param has no default value
```

# Note
Found out that you need to explicitly pass `None` as the default, otherwise it will crash with above.

### What you think should happen instead

`month` should be passed to the dag

### Operating System

mac osx

### Deployment

Virtualenv installation

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by potiuk** [comment]
Marked it as good first issue. Hopefully somoene will fix it soon (but it's up for grabs for you if you would like to fix it ). PRs are most welcome to fix it.

**Comment by amoghrajesh** [comment]
@potiuk @aamster I can take this up if it is not already taken. Would be more than happy to have this fixed.

**Comment by Taragolis** [comment]
> Found out that you need to explicitly pass None as the default, otherwise it will crash with above.

For me it make sense. If no default value provided than parameter is a mandatory, and otherwise if default parameter provided than parameter is optional

**Comment by hussein-awala** [comment]
I agree with @Taragolis, we should check if the required parameters (with no default value) are provided or not, and fail the dag run if they are not provided.
I'm working on a #28968 to deprecate dag run conf and support providing the params directly, and this exception will be raised before creating the dag run ([code](https://github.com/apache/airflow/pull/29174/files#diff-62c8e300ee91e0d59f81e0ea5d30834f04db71ae74f2e155a10b51056b00b59bR2567))

**Comment by hussein-awala** [comment]
It seems like currently we don't support required params, to fix this, I think we need to differentiate between parsing the params during the dag parsing/creation and parsing the params during the dag run execution.

During the dag parsing, we should not raise an exception if the default is not provided because this param is defined as required param. But during the dag run execution, if the param is required and the value is not provided, we should raise an exception to fail the dag run or the task if the params are defined at the task level.

I don't think this is a good first issue.

**Comment by potiuk** [comment]
Sure. Removed the label then

**Comment by jscheffl** [comment]
There had been a couple of PRs by me, for example #31301 and #34248 which handle optional fields and how Params are handled for this. Just got note of this bug report and must state after a manual re-test: Can not be re-produced. Did not explicitly test but might have been resolved in 2.8.0 already. Therefore closing.
